itests/kit/node_unmanaged.go

```diff
@@ -464,332 +685,424 @@ func (tm *TestUnmanagedMiner) SubmitProveCommit(
+		var err error
+		msgReturn, err = tm.SubmitMessage(&miner14.ProveCommitSectors3Params{
+			SectorActivations:        activations,
+			SectorProofs:             sectorProofs,
+			RequireActivationSuccess: requireActivationSuccess,
+		}, 0, builtin.MethodsMiner.ProveCommitSectors3)
+		req.NoError(err)
+		req.True(msgReturn.Receipt.ExitCode.IsSuccess())
 	}
 
-	tm.t.Logf("Miner %s: WindowPoST(%d): SectorPartition: %+v", tm.ActorAddr, sectorNumber, sp)
-
-	// Calculate the start of the period, adjusting if the current deadline has passed
-	periodStart := di.PeriodStart
-	// calculate current deadline index because it won't be reliable from state until the first
-	// challenge window cron tick after first sector onboarded
-	currentDeadlineIdx := uint64(math.Abs(float64((di.CurrentEpoch - di.PeriodStart) / di.WPoStChallengeWindow)))
-	if di.PeriodStart < di.CurrentEpoch && sp.Deadline <= currentDeadlineIdx {
-		// If the deadline has passed in the current proving period, calculate for the next period
-		// Note that di.Open may be > di.CurrentEpoch if the miner has just been enrolled in cron so
-		// their deadlines haven't started rolling yet
-		periodStart += di.WPoStProvingPeriod
+	var returnValue miner14.BatchReturn
+	req.NoError(returnValue.UnmarshalCBOR(bytes.NewReader(msgReturn.Receipt.Return)))
+	exitCodes := exitCodesFromBatchReturn(returnValue)
+
+	for i, sector := range sectors {
+		si, err := tm.FullNode.StateSectorGetInfo(tm.ctx, tm.ActorAddr, sector.SectorNumber, msgReturn.TipSet)
+		tm.log("SectorOnChainInfo for sector %d w/ exit code %s: %+v", sector.SectorNumber, exitCodes[i], si)
+		if exitCodes[i] == exitcode.Ok {
+			req.NoError(err)
+			req.Equal(si.SectorNumber, sector.SectorNumber)
+			req.Equal(si.Activation, msgReturn.Height-1)
+		} else {
+			req.Nil(si, "sector should not be on chain")
+		}
+		if proofType.IsNonInteractive() {
+			// NI-PoRep lets us determine the deadline, so we can check that it's set correctly
+			sp, err := tm.FullNode.StateSectorPartition(tm.ctx, tm.ActorAddr, sector.SectorNumber, msgReturn.TipSet)
+			if exitCodes[i] == exitcode.Ok {
+				req.NoError(err)
+				req.Equal(provingDeadline, sp.Deadline)
+			} else {
+				req.ErrorContains(err, fmt.Sprintf("sector %d not due at any deadline", sector.SectorNumber))
+			}
+		}
 	}
 
-	// Calculate the exact epoch when proving should occur
-	provingEpoch := periodStart + di.WPoStChallengeWindow*abi.ChainEpoch(sp.Deadline)
-
-	tm.t.Logf("Miner %s: WindowPoST(%d): next ProvingEpoch: %d", tm.ActorAddr, sectorNumber, provingEpoch)
-
-	return di.CurrentEpoch, provingEpoch, nil
+	return exitCodes
 }
 
-func (tm *TestUnmanagedMiner) wdPostLoop(
-	pctx context.Context,
-	sectorNumber abi.SectorNumber,
-	sealedCid cid.Cid,
-	sealedPath,
-	cacheDir string,
-) (chan WindowPostResp, context.CancelFunc) {
-
-	ctx, cancelFn := context.WithCancel(pctx)
-	respCh := make(chan WindowPostResp, 1)
+func (tm *TestUnmanagedMiner) wdPostLoop() {
+	if tm.runningWdPostLoop {
```

Comment:
There should only ever be one wdPostLoop per miner and so this should never happen/something has gone wrong if this happens. I think we should get rid of this variable.

Reply:
The current code allows for multiple calls to `OnboardSectors`, although it's not something I've exercised yet, but everything is in place for it to _just work_. At the end of onboarding it runs `tm.wdPostLoop()` and that should be a noop if it doesn't need to start it.